polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -0,0 +1,285 @@
+  * a client sends an edit to the server, it includes the latest global version it knows about, and its local version,
+  * which it then increments. When the server broadcasts an edit, it sends a message to each subscriber containing
+  * the global version of the edit, and the local version to which it applies for that client. Then, the client can
+  * transform the edit (if necessary) to its current local version, and it will be in a consistent state with the
+  * global version.
+  *
+  * Similarly, the server can do part of the transformation work – when it receives an edit that acts upon global version
+  * X, but it knows that client A is currently on at least version X + 3, it can transform the edit to act upon
+  * version X + 3 and send that transformed edit to client A (which will also apply any necessary transformations locally).
+  *
+  * Thus, the SharedNotebook must track its global version, the current notebook, and low watermark of the highest global version
+  * that all clients have acknowledged. It must also keep a history buffer of all edits between the low watermark and
+  * the current global version. Clients can periodically send a status message to the server indicating the current-known
+  * global version and the current local version, in order to allow the server to discard edit history.
+  *
+  * Clients must track the currently known global version, which updates whenever the client receives a foreign edit,
+  * and the highest local version which has been acknowledged by the server (either by sending an edit based on it,
+  * or through an acknowledgement message). It has to keep its local edit history between that known-acknowledged
+  * version and its current version, to allow rebasing new foreign edits.
+  *
+  * It sounds a lot more complicated than it really is.
+  */
+trait SharedNotebook[F[_]] {
+
+  def path: String
+
+  /**
+    * Open a reference to this shared notebook.
+    *
+    * @param name A string identifying who is opening the notebook (i.e. their username or email)
+    * @param oq   A queue to which messages to the subscriber (such as notifications of foreign updates) will be submitted
+    * @return A [[NotebookRef]] which the caller can use to send updates to the shared notebook
+    */
+  def open(name: String, oq: Enqueue[F, Message]): F[NotebookRef[F]]
+
+  def versions: Stream[F, (Int, Notebook)]
+}
+
+
+class IOSharedNotebook(
+  val path: String,
+  ref: SignallingRef[IO, (GlobalVersion, Notebook)],            // the Int is the global version, which can wrap around back to zero if necessary
+  kernelRef: Ref[IO, Option[Kernel[IO]]],
+  updates: Queue[IO, Option[(SubscriberId, NotebookUpdate, Deferred[IO, GlobalVersion])]],   // the canonical series of edits
+  updatesTopic: Topic[IO, Option[(GlobalVersion, SubscriberId, NotebookUpdate)]],  // a subscribe-able channel for watching updates,
```

Comment:
It looks like the only place we create `updates` and `updatesTopic` is in `IOSharedNotebook.apply` - in which case couldn't they be moved out of the constructor and created internally? That would ensure that we don't get any unexpected updates from other sources - unless the intention is to enable that?

Reply:
The main reason it's in the constructor is that the constructors for `Queue` and `Topic` (and `Ref` and `SignallingRef`...sigh) all return i.e. `F[Ref[F, A]]`. So they're inside IO at this point. I can either make the `SharedNotebook` do the same, and create those things in the constructor, or I can make it a private field and get it with `unsafeRunSync`. I think it's better practice to do the former (though the latter happens sometimes too...)

Reply:
Ah, I see - I didn't notice that they were inside their own `IO`s when constructed.

It's hard to wrap my head around how many `IO`s there are all over the place (and how many nestings of IOs there are..._